### PR TITLE
Replace usage of `grep -P` with `perl` in `run_standalone_tests.sh`

### LINKED ISSUE #19938: `grep: Invalid option -- P` when running `./tests/run_standalone_tests.sh` on macOS
### Bug description

BSD `grep` no longer supports the `-- P` option on macOS.

One fix would be to use GNU `grep` instead, as `ggrep`.

### What version are you seeing the problem on?

master

### How to reproduce the bug

_No response_

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @borda

### PR DESCRIPTION
**Pull request by Peiffap**
## What does this PR do?
Fixes #19938, a "bug" where on macOS, which does not use GNU `grep` by default and does not recognize the `-P` flag, `run_standalone_tests.sh` errors out (earlier than expected).

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19942.org.readthedocs.build/en/19942/

### CODE REVIEW

**Review by awaelchli — COMMENTED** [review]
Thanks!

File: tests/run_standalone_tests.sh [hunk 1]
```diff
@@ -26,18 +26,17 @@ export PL_RUN_STANDALONE_TESTS=1
...
 test_path=$1
 printf "source path: $test_path\n"
 
 # collect all tests with parametrization based filtering with PL_RUN_STANDALONE_TESTS
 standalone_tests=$(python3 -m pytest $test_path -q --collect-only --pythonwarnings ignore)
-printf "Collected tests: \n $standalone_tests"
+printf "Collected tests: \n $standalone_tests\n"
 # match only lines with tests
-parametrizations=$(grep -oP '\S+::test_\S+' <<< "$standalone_tests")
+parametrizations=$(perl -nle'print $& while m{\S+::test_\S+}g' <<< "$standalone_tests")
```
Comment (awaelchli) [inline tests/run_standalone_tests.sh:37]: This looks odd, shouldn't there be a space here between argument and value?
```suggestion
parametrizations=$(perl -nle 'print $& while m{\S+::test_\S+}g' <<< "$standalone_tests")
```

File: tests/run_standalone_tests.sh [hunk 2]
```diff
@@ -47,7 +46,7 @@ function show_batched_output {
   if [ -f standalone_test_output.txt ]; then  # if exists
     cat standalone_test_output.txt
     # heuristic: stop if there's mentions of errors. this can prevent false negatives when only some of the ranks fail
-    if grep -iE 'error|exception|traceback|failed' standalone_test_output.txt | grep -vE 'on_exception|xfailed' | grep -qv -f testnames.txt; then
+    if perl -nle'print if /error|(?<!(?-i)on_)exception|traceback|(?<!(?-i)x)failed/i' standalone_test_output.txt | grep -qv -f testnames.txt; then
```
Comment (Peiffap) [inline tests/run_standalone_tests.sh:49]:
```suggestion
    if perl -nle 'print if /error|(?<!(?-i)on_)exception|traceback|(?<!(?-i)x)failed/i' standalone_test_output.txt | grep -qv -f testnames.txt; then
```

File: tests/run_standalone_tests.sh (hunk 1 above)
Comment (Peiffap) [reply tests/run_standalone_tests.sh]: Good catch, thanks. I did the same below for the other call.

**Review by awaelchli — APPROVED** [review]

**Review by lantiga — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Pull request merged**